src/cirrus/lib/task.py

```diff
@@ -0,0 +1,208 @@
+from abc import ABC, abstractmethod #abstractclassmethod
+import argparse
+from copy import deepcopy
+import json
+import logging
+from os import makedirs
+import os.path as op
+from shutil import rmtree
+import sys
+from tempfile import mkdtemp
+from typing import Dict, List, Optional
+
+from boto3utils import s3
+
+from cirrus.lib.logging import get_task_logger
+from cirrus.lib.process_payload import ProcessPayload
+from cirrus.lib.transfer import download_item_assets, upload_item_assets
+
+
+class Task(ABC):
+
+    _name = 'task'
+    _description = 'A task for doing things'
+    _version = '0.1.0'
+
+    def __init__(self: "Task", payload: Dict, local: Optional[bool]=None, workdir: Optional[str]=None):
+        # parse event
+        payload = ProcessPayload.from_event(payload)
+
+        # TODO - use PySTAC here
+        self.original_items = payload.pop('features')
+        self.items = deepcopy(self.original_items)
+
+        self.process_definition = payload.pop('process')
+        self.remaining_payload = dict(payload)
+
+        # set up logger
+        self.logger = get_task_logger(f"task.{self._name}", payload=payload)
+        
+        # local mode? 
+        self._local = local
+
+        # create temporary work directory if workdir is None
+        self._workdir = workdir
+        if workdir is None:
+            self._workdir = mkdtemp()
+            self._tmpworkdir = True
+        else:
+            self._workdir = workdir
+            self._tmpworkdir = False
+            makedirs(self._workdir, exist_ok=True)
+
+        #self.validate()
+
+    @property
+    def parameters(self):
+        return self.process_definition['tasks'].get(self._name, {})
+
+    @property
+    def output_options(self):
+        return self.process_definition.get('output_options', {})
+
+    def validate(self):
+        # put validation logic on input Items and process definition here
+        pass
+
+    def download_assets(self, assets: Optional[List[str]]=None):
+        """Download provided asset keys for all items in payload. Assets are saved in workdir in a
+           directory named by the Item ID
+
+        Args:
+            assets (Optional[List[str]], optional): List of asset keys to download. Defaults to all assets.
+        """
+        for i, item in enumerate(self.items):
+            outdir = op.join(self._workdir, item['id'])
```

Comment:
It seems like this should probably be like `op.join(self._workdir, item['collection'], item['id'])`.

As an aside, I highly recommend switching from `os.path` to using `pathlib.Path` objects.

Reply:
Will switch to pathlib